File: parent/wymiwyg/jetty-osgi-httpservice/core/src/main/java/org/wymiwyg/jetty/httpservice/Activator.java
/*
 * Copyright  2002-2006 WYMIWYG (http://wymiwyg.org)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package org.wymiwyg.jetty.httpservice;

import java.util.HashSet;
import java.util.Set;
import javax.net.ssl.SSLContext;
import org.apache.felix.scr.annotations.Component;
import org.apache.felix.scr.annotations.Reference;
import org.apache.felix.scr.annotations.ReferenceCardinality;
import org.apache.felix.scr.annotations.ReferencePolicy;
import org.eclipse.equinox.http.servlet.HttpServiceServlet;

import org.eclipse.jetty.server.Connector;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.nio.SelectChannelConnector;
import org.eclipse.jetty.server.ssl.SslConnector;
import org.eclipse.jetty.server.ssl.SslSocketConnector;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;
import org.osgi.framework.BundleContext;
import org.osgi.service.component.ComponentContext;

/**
 * @author reto
 */
@Component(immediate = true)
@Reference(name="sslContext",
		cardinality=ReferenceCardinality.OPTIONAL_MULTIPLE,
		policy=ReferencePolicy.DYNAMIC,
		referenceInterface=SSLContext.class)
public class Activator {


	public static final String CONTEXT_PROPERTY_HTTP_PORT = "org.osgi.service.http.port";
	public static final String CONTEXT_PROPERTY_HTTP_PORT_SECURE = "org.osgi.service.http.port.secure";
	public static final String CONTEXT_PROPERTY_HTTPS_ENABLED = "org.osgi.service.http.secure.enabled";
	public static final String CONTEXT_PROPERTY_HTTP_ENABLED = "org.osgi.service.http.enabled";
	public static final String CONTEXT_PROPERTY_KEYSTORE_PASSWORD = "org.wymiwyg.jetty.httpservice.https.keystore.password";
	public static final String CONTEXT_PROPERTY_KEYSTORE_PATH = "org.wymiwyg.jetty.httpservice.https.keystore.path";
	public static final String CONTEXT_PROPERTY_KEYSTORE_TYPE = "org.wymiwyg.jetty.httpservice.https.keystore.type";
	/**
	 * Context property, one of "none", "want" or "need"
	 */
	public static final String CONTEXT_PROPERTY_CLIENTAUTH = "org.wymiwyg.jetty.httpservice.clientauth";
	private Server server;
	private ComponentContext context;
	private Set<SSLContext> sslContexts = new HashSet<SSLContext>();

	protected void activate(ComponentContext context) throws Exception {
		this.context = context;
		start();
	}
	
	private void start() throws Exception {
		server = new Server();
		BundleContext bundleContext = context.getBundleContext();
		ServletContextHandler servletContext = new ServletContextHandler(ServletContextHandler.SESSIONS);
		servletContext.setContextPath("/");
		server.setHandler(servletContext);
		servletContext.addServlet(new ServletHolder(new HttpServiceServlet()), "/*");

		if (isHttpEnabled(bundleContext)) {
			Connector connector = new SelectChannelConnector();
			connector.setPort(getHttpPort(bundleContext));
			server.addConnector(connector);
		}
		if (isHttpsEnabled(bundleContext)) {
			SslConnector sslConnector = new SslSocketConnector();
			sslConnector.setKeyPassword(getKeyPassword(bundleContext));
			String keyStorePath = bundleContext.getProperty(CONTEXT_PROPERTY_KEYSTORE_PATH);
			if (keyStorePath != null) {
				sslConnector.setKeystore(keyStorePath);
			}
			String keyStoreType = bundleContext.getProperty(CONTEXT_PROPERTY_KEYSTORE_TYPE);
			if (keyStoreType != null) {
				sslConnector.setKeystoreType(keyStoreType);
			}
			if(sslContexts.size() > 0) {
				sslConnector.setSslContext(sslContexts.iterator().next());
			}
			sslConnector.setPort(getHttpsPort(bundleContext));
			String clientAuth = bundleContext.getProperty(CONTEXT_PROPERTY_CLIENTAUTH);
			if ("want".equals(clientAuth)) {
				sslConnector.setWantClientAuth(true);
			} else {
				if ("need".equals(clientAuth)) {
					sslConnector.setNeedClientAuth(true);
				}
			}
			server.addConnector(sslConnector);
		}
		server.start();
	}

	private void stop() throws Exception {
		server.stop();
	}
	private void restart() throws Exception {
		if (server != null) {
			stop();
			start();
		}
	}
	protected void deactivate(ComponentContext context) throws Exception {
		stop();
		server = null;
	}
	
	public void bindSslContext(SSLContext sslContext) {
		sslContexts.add(sslContext);
		if (sslContexts.size() == 1) {
			try {
				restart();
			} catch (Exception ex) {
				throw new RuntimeException(ex);
			}
		}
	}
	
	public void unbindSslContext(SSLContext sslContext) throws Exception {
		sslContexts.remove(sslContext);
		restart();
	}

	private boolean isHttpEnabled(BundleContext bundleContext) {
		final String property = bundleContext.getProperty(CONTEXT_PROPERTY_HTTP_ENABLED);
		if (property != null) {
			return Boolean.parseBoolean(property);
		} else {
			return true;
		}
	}

	private boolean isHttpsEnabled(BundleContext bundleContext) {
		final String property = bundleContext.getProperty(CONTEXT_PROPERTY_HTTPS_ENABLED);
		if (property != null) {
			return Boolean.parseBoolean(property);
		} else {
			return false;
		}
	}

	private int getHttpPort(BundleContext bundleContext) {
		final String property = bundleContext.getProperty(CONTEXT_PROPERTY_HTTP_PORT);
		if (property != null) {
			return Integer.parseInt(property);
		} else {
			return 8080;
		}
	}
	
	private int getHttpsPort(BundleContext bundleContext) {
		final String property = bundleContext.getProperty(CONTEXT_PROPERTY_HTTP_PORT_SECURE);
		if (property != null) {
			return Integer.parseInt(property);
		} else {
			return 8443;
		}
	}

	private String getKeyPassword(BundleContext bundleContext) {
		final String property = bundleContext.getProperty(CONTEXT_PROPERTY_KEYSTORE_PASSWORD);
		if (property != null) {
			return property;
		} else {
			return "password";
		}
	}
}

File: parent/platform.security.foafssl/core/src/main/scala/org/apache/clerezza/foafssl/ssl/Activator.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.clerezza.foafssl.ssl

import java.security.KeyStore
import java.util._
import java.io._
import javax.net.ssl.SSLContext
import org.jsslutils.keystores.KeyStoreLoader
import org.jsslutils.sslcontext.X509SSLContextFactory
import org.osgi.framework.BundleContext;
import org.osgi.service.component.ComponentContext
import org.wymiwyg.jetty.httpservice.{Activator => ServiceActivator}

object Activator {



	def getKeyStoreType(context: BundleContext) = {
		val property = context.getProperty(ServiceActivator.CONTEXT_PROPERTY_KEYSTORE_TYPE);
		if (property != null) {
			property;
		} else {
			"JKS"
		}
	}

	def getKeyStorePath(context: BundleContext) = {
		val property = context.getProperty(ServiceActivator.CONTEXT_PROPERTY_KEYSTORE_PATH);
		if (property != null) {
			property;
		} else {
			new File(new File(System.getProperty("user.home")), ".keystore").getAbsolutePath
		}
	}

	//todo: should check what apps can get access to this. Is it properly protected?
	def getKeyStorePassword(context: BundleContext) = {
		val property = context.getProperty(ServiceActivator.CONTEXT_PROPERTY_KEYSTORE_PASSWORD);
		if (property != null) {
			property;
		} else {
			"password";
		}
	}

	def getServerCertKeyStore(bundleContext: BundleContext): KeyStore = {
		val keyStoreLoader = new KeyStoreLoader
		keyStoreLoader.setKeyStoreType(getKeyStoreType(bundleContext))
		keyStoreLoader.setKeyStorePath(getKeyStorePath(bundleContext))
		keyStoreLoader.setKeyStorePassword(getKeyStorePassword(bundleContext))
      return keyStoreLoader.loadKeyStore();
    }

}

class Activator() {
	import Activator._

	var x509TrustManagerWrapperService: X509TrustManagerWrapperService = null;
	protected def bindX509TrustManagerWrapperService(s: X509TrustManagerWrapperService)  = {
		x509TrustManagerWrapperService = s
	}
	protected def unbindX509TrustManagerWrapperService(s: X509TrustManagerWrapperService)  = {
		x509TrustManagerWrapperService = null
	}
	//registering the SSLContext service should start the https service -- org.wymiwyg.jetty.httpservice.Activator does for example
	protected def activate(context: ComponentContext) = {
    try{
      val bundleContext = context.getBundleContext
      //TODO set jvm default ca-store
      val http = bundleContext.getProperty("org.osgi.service.http.secure.enabled")
      if (http!=null && "true".equals(http)) {
        val sslContextFactory = new X509SSLContextFactory(
          getServerCertKeyStore(bundleContext),
          getKeyStorePassword(bundleContext),
          getServerCertKeyStore(bundleContext)); //getCaKeyStore());
        sslContextFactory
          .setTrustManagerWrapper(x509TrustManagerWrapperService);
        val sslContext = sslContextFactory.buildSSLContext("TLS")
        x509TrustManagerWrapperService.setSslContext(sslContext)

        bundleContext.registerService(classOf[SSLContext].getName, sslContext, new Properties())
        println("Registered SSLContext+")
      }
    }
    catch{
      case e : Exception => println("unable to activate FOAF+SSL")
    }
	}
	
	
	

}


